paddle/fluid/framework/program_processing.cc

```diff
@@ -0,0 +1,135 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/framework/program_processing.h"
+#include "paddle/fluid/framework/block_desc.h"
+
+namespace paddle {
+namespace framework {
+
+void ProgramProcessor::GetInputsOutputsInBlock(
+    const BlockDesc &current_block, std::set<std::string> *inner_inputs,
+    std::set<std::string> *inner_outputs) {
+  /* Find inputs and outputs in current control flow block.
+  :param current_block: Current control flow block.
+  :param inner_inputs: Input var vector of ops in current block.
+  :param inner_outputs: Output var vector of ops in current block. */
+
+  // Step1: update inner_inputs and inner_outputs
+  // NOTE: Here assumes that all variables are input or output of Ops,
+
+  std::set<std::string> removed_inner_inputs;
+
+  for (OpDesc *op : current_block.AllOps()) {
+    for (auto iname : op->InputNames()) {
+      for (auto in_var_name : op->Input(iname)) {
+        VLOG(3) << "in_var_name:" << in_var_name;
+        if (inner_outputs->find(in_var_name) == inner_outputs->end()) {
+          VLOG(3) << "insert inner_inputs_name:" << in_var_name;
+          inner_inputs->insert(in_var_name);
+        }
+      }
+    }
+
+    for (auto oname : op->OutputNames()) {
+      for (auto out_var_name : op->Output(oname)) {
+        VLOG(3) << "insert out_var_name:" << out_var_name;
+        inner_outputs->insert(out_var_name);
+      }
+    }
+  }
+
+  // Step2: Remove variable created in current control flow block.
+  BlockDesc *parent_block = current_block.ParentBlock();
+  VarDesc *current_block_var;
```

Comment:
Dangerous code: you should write
`VarDesc *current_block_var = nullptr;`
in your case because C++ doesn't initialize the value, and you have a condition `&&current_block_var` at line 63.

Reply:
Done.